File: td/telegram/Log.cpp
//
// Copyright Aliaksei Levin (levlam@telegram.org), Arseny Smirnov (arseny30@gmail.com) 2014-2025
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
#include "td/telegram/Log.h"

#include "td/telegram/Client.h"
#include "td/telegram/Logging.h"
#include "td/telegram/td_api.h"

#include "td/utils/common.h"

#include <mutex>

namespace td {

static std::mutex log_mutex;
static string log_file_path;
static int64 max_log_file_size = 10 << 20;
static Log::FatalErrorCallbackPtr fatal_error_callback;

static void fatal_error_callback_wrapper(int verbosity_level, const char *message) {
  if (verbosity_level == 0) {
    auto callback = fatal_error_callback;
    if (callback != nullptr) {
      callback(message);
    }
  }
}

bool Log::set_file_path(string file_path) {
  std::lock_guard<std::mutex> lock(log_mutex);
  if (file_path.empty()) {
    log_file_path.clear();
    return Logging::set_current_stream(td_api::make_object<td_api::logStreamDefault>()).is_ok();
  }

  if (Logging::set_current_stream(td_api::make_object<td_api::logStreamFile>(file_path, max_log_file_size, true))
          .is_ok()) {
    log_file_path = std::move(file_path);
    return true;
  }

  return false;
}

void Log::set_max_file_size(int64 max_file_size) {
  std::lock_guard<std::mutex> lock(log_mutex);
  max_log_file_size = max(max_file_size, static_cast<int64>(1));
  Logging::set_current_stream(td_api::make_object<td_api::logStreamFile>(log_file_path, max_log_file_size, true))
      .ignore();
}

void Log::set_verbosity_level(int new_verbosity_level) {
  std::lock_guard<std::mutex> lock(log_mutex);
  Logging::set_verbosity_level(new_verbosity_level).ignore();
}

void Log::set_fatal_error_callback(FatalErrorCallbackPtr callback) {
  std::lock_guard<std::mutex> lock(log_mutex);
  if (callback == nullptr) {
    ClientManager::set_log_message_callback(0, nullptr);
    fatal_error_callback = nullptr;
  } else {
    fatal_error_callback = callback;
    ClientManager::set_log_message_callback(0, fatal_error_callback_wrapper);
  }
}

}  // namespace td

File: example/java/org/drinkless/tdlib/Log.java
//
// Copyright Aliaksei Levin (levlam@telegram.org), Arseny Smirnov (arseny30@gmail.com) 2014-2020
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//
package org.drinkless.tdlib;

/**
 * Class used for managing internal TDLib logging.
 * Use TdApi.*Log* methods instead.
 */
public final class Log {
    /**
     * Changes TDLib log verbosity.
     *
     * @deprecated As of TDLib 1.4.0 in favor of {@link TdApi.SetLogVerbosityLevel}, to be removed in the future.
     * @param verbosityLevel New value of log verbosity level. Must be non-negative.
     *                       Value 0 corresponds to fatal errors,
     *                       value 1 corresponds to java.util.logging.Level.SEVERE,
     *                       value 2 corresponds to java.util.logging.Level.WARNING,
     *                       value 3 corresponds to java.util.logging.Level.INFO,
     *                       value 4 corresponds to java.util.logging.Level.FINE,
     *                       value 5 corresponds to java.util.logging.Level.FINER,
     *                       value greater than 5 can be used to enable even more logging.
     *                       Default value of the log verbosity level is 5.
     */
    @Deprecated
    public static native void setVerbosityLevel(int verbosityLevel);

    /**
     * Sets file path for writing TDLib internal log. By default TDLib writes logs to the System.err.
     * Use this method to write the log to a file instead.
     *
     * @deprecated As of TDLib 1.4.0 in favor of {@link TdApi.SetLogStream}, to be removed in the future.
     * @param filePath Path to a file for writing TDLib internal log. Use an empty path to
     *                 switch back to logging to the System.err.
     * @return whether opening the log file succeeded.
     */
    @Deprecated
    public static native boolean setFilePath(String filePath);

    /**
     * Changes the maximum size of TDLib log file.
     *
     * @deprecated As of TDLib 1.4.0 in favor of {@link TdApi.SetLogStream}, to be removed in the future.
     * @param maxFileSize The maximum size of the file to where the internal TDLib log is written
     *                    before the file will be auto-rotated. Must be positive. Defaults to 10 MB.
     */
    @Deprecated
    public static native void setMaxFileSize(long maxFileSize);

    /**
     * This function is called from the JNI when a fatal error happens to provide a better error message.
     * The function does not return.
     *
     * @param errorMessage Error message.
     */
    private static void onFatalError(String errorMessage) {
        class ThrowError implements Runnable {
            private ThrowError(String errorMessage) {
                this.errorMessage = errorMessage;
            }

            @Override
            public void run() {
                throw new RuntimeException("TDLib fatal error: " + errorMessage);
            }

            private final String errorMessage;
        }

        new Thread(new ThrowError(errorMessage), "TDLib fatal error thread").start();
        while (true) {
            try {
                Thread.sleep(1000);     // milliseconds
            } catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
